### PR TITLE
Import Stim detector error models for Pauli-frame sampling

### LINKED ISSUE #728: Import Stim detector error models for Pauli-frame-style sampling
This issue is part of **unitaryHACK26**. You have to be [registered](https://unitaryhack.dev/register/) to complete this issue.
Learn more about the PR submission process [here](https://unitaryhack.dev/hacker-guide/) and about unitaryHACK rules [here](https://unitary.design/rules/)!

**A note about AI Slop:** while we are open to collaboration with LLMs for unitaryHACK, fully AI-generated PRs are not acceptable. It is at the maintainers' discretion whether or not LLM-generated PRs are the right fit for the issues, and those that appear fully AI-generated may be immediately rejected. Read unitaryHACK's full AI Policy [here](https://unitaryhack.dev/ai-guide/).

Note from the maintainer: If you have not contributed before to open source projects, no AI agent use is permitted for this issue.

---

Many quantum error-correction workflows use Stim detector error model (`.dem`)
files as an interchange format. QuantumClifford has fast Pauli-frame simulation
machinery, but there is currently no convenient way to ingest an existing `.dem`
file and turn it into a QuantumClifford circuit. Importing `.dem` files as
circuits would let users compare Stim-generated models with QuantumClifford
decoders and analysis tools while continuing to use the existing
`pftrajectories` backend for sampling.

Relevant links:

- QuantumClifford repository: https://github.com/QuantumSavory/QuantumClifford.jl
- Pauli-frame implementation: https://github.com/QuantumSavory/QuantumClifford.jl/blob/main/src/pauli_frames.jl
- Existing noisy circuit operations: https://github.com/QuantumSavory/QuantumClifford.jl/blob/main/src/noise.jl
- Stim detector error model format: https://github.com/quantumlib/Stim/blob/main/doc/file_format_dem_detector_error_model.md

## Desired user-facing behavior

A user should be able to load a `.dem` file and receive a corresponding
QuantumClifford circuit that can be passed directly to `pftrajectories`:

```julia
using QuantumClifford

circuit = read_detector_error_model("surface_code.dem")
frames = pftrajectories(circuit; trajectories=10_000)

measurements(frames)
```

The exact function name can change, but the output should be a circuit-like
object, a vector of operations, that is accepted by the
existing `pftrajectories` methods:

```julia
circuit = detector_error_model_circuit("surface_code.dem")
frames = pftrajectories(circuit; trajectories=10_000)
measurements(frames)
```

The important requirement is that sampling remains implemented by
`pftrajectories`. This bounty should not introduce a separate sampling backend.

## Suggested implementation

Implement a parser for the useful subset of Stim detector error model syntax:

- `error(p) D... L...`
- `detector ...`
- `logical_observable ...`
- `shift_detectors ...`
- `repeat K { ... }`
- comments and blank lines

Coordinates on `detector` or `shift_detectors` lines can be parsed and stored if
convenient, but it is acceptable to ignore them initially as long as the detector
index shifts are handled correctly.

The parser should lower the `.dem` file to operations that `pftrajectories` can
execute. If the existing gate set is not expressive enough, add a small number of
new circuit operations for detector error model semantics. For example, an
operation like this may be useful:

```julia
DetectorError(probability, detector_bits, logical_bits)
```

When applied to a `PauliFrame`, this operation would sample one independent
Bernoulli event per trajectory and XOR the corresponding detector and logical
bits into the frame's measurement storage. A no-op declaration operation may also
be needed so explicitly declared detectors or logical observables with no errors
still allocate output columns.

The semantics should match Stim's `.dem` documentation: each `error(p)` line is
an independent error mechanism, and when it occurs it flips the listed detector
and logical targets. The implementation should express that behavior through
QuantumClifford circuit operations, then rely on `pftrajectories` to run many
trajectories efficiently.

## Minimal example

Given this file:

```text
detector D0
detector D1
logical_observable L0
error(0.1) D0
error(0.2) D0 D1 L0
```

the importer should produce a model with two detectors, one logical observable,
and two independent error mechanisms. The returned circuit should be accepted by
`pftrajectories`; sampling that circuit many times should produce detector and
logical outcome frequencies consistent with those probabilities.

## In scope

- A documented circuit representation for imported detector error models.
- A parser for common Stim `.dem` files, including nested `repeat` blocks and
  `shift_detectors`.
- Any small new `AbstractOperation` subtypes needed to represent `.dem`
  semantics in circuits.
- `apply!`/`affectedbits` support needed for those operations to work with
  `PauliFrame` and `pftrajectories`.
- Tests comparing small hand-written `.dem` files against known expected
  detector/logical distributions.
- Optional statistical comparison against Stim's own sampler when Stim is
  available.
- Clear error messages for unsupported syntax.

## Out of scope

- Parsing Stim circuit (`.stim`) files.
- Decoding the sampled detector data.
- A new sampling engine separate from `pftrajectories`.
- Exact bit-for-bit matching of Stim's random number generator.
- Full preservation of coordinate metadata, unless this is straightforward.

## Acceptance criteria

- `read_detector_error_model(path)` or an equivalent API imports a `.dem` file
  and returns a circuit-like object accepted by `pftrajectories`.
- The parser handles `error`, `detector`, `logical_observable`,
  `shift_detectors`, and `repeat` for representative files.
- New `.dem`-specific operations, if added, implement the `affectedbits` and
  `apply!` methods needed by `pftrajectories`.
- `pftrajectories(imported_circuit; trajectories=...)` returns detector and
  logical bit arrays with documented dimensions.
- Statistical tests validate the imported circuit on small models.
- Documentation includes a short imported `.dem` example and the expected output
  shape.

### PR DESCRIPTION
**Pull request by GrovyleX**
Closes #728

Adds a way to load Stim `.dem` files into QuantumClifford and sample them with the existing `pftrajectories` backend. No separate sampler - each `error(p)` line becomes one `DetectorError` op that flips detector/observable measurement columns with probability `p`.

### What changed

- added `src/dem.jl` to import `.dem` files via `read_detector_error_model` / `parse_detector_error_model`
- new ops `DetectorError` and `DeclareMeasurementBits` so `pftrajectories` can sample detector/observable flips directly
- parser covers the usual Stim subset (`error`, `detector`, `logical_observable`, `shift_detectors`, `repeat`, comments/tags); coords and `^` separators are ignored for sampling
- docs + tests (`test/test_dem.jl`, including a statistical sanity check on flip rates)
- CHANGELOG + version bump to 0.11.5-dev

### Output layout

After `pftrajectories`, `measurements(frames)` is `trajectories × (num_detectors + num_observables)`. Detectors are the first columns, observables after. Values are relative to the noiseless reference run, same as other Pauli-frame sims.

### Notes

- `repeat` blocks expand eagerly (same idea as Stim, but a huge repeat count means a huge op vector)
- mechanisms with more than 16 targets fall back to the non-compactified path (warning from `pftrajectories`) - should not matter for typical graphlike DEMs

### Test plan

- ran `test/test_dem.jl` via TestItemRunner locally (34 tests)
- checked the minimal issue example: empirical freqs ~ [0.26, 0.2, 0.2] vs analytic values
- smoke-tested `repeat` / `shift_detectors` / nested blocks / file read

### AI usage disclosure

I used Claude (Claude Code) for some planning and early implementation scaffolding on this PR. I reviewed the full diff myself, adjusted code to match repo style, fixed parser edge cases, wrote and ran the tests (including statistical checks).

I have read and understood the submitted code, and I ran the tests locally before opening this PR.

- [x] The code is properly formatted and commented.
- [x] Substantial new functionality is documented within the docs.
- [x] All new functionality is tested.
- [ ] All of the automated tests on github pass.
- [x] We recently started enforcing formatting checks. If formatting issues are reported in the new code you have written, please correct them. <small>There will be plenty of old code that is flagged as we are slowly transitioning to enforced formatting. Please do not worry about or address older formatting issues -- keep your PR just focused on your planned contribution.</small>

If you are submitting for a bug bounty:
- [x] I have read and followed the [AI usage and disclosure policy](https://github.com/QuantumSavory/.github/blob/main/BUG_BOUNTIES.md)

### CODE REVIEW

**Comment by Krastanov** [comment]
We do not accept LLM-generated PRs by unknown developers for our bounties.

**Comment by Krastanov** [comment]
Apologies, this was trigger happy on me -- I did not notice that you had actually done the checkboxes and that this is not completely bot-written.

**Comment by GrovyleX** [comment]
I am glad you noticed. I will continue working with this PR. 1 CI check is failing. I will try fixing that. Apart from that will working on improving the PR... Is possible give some review so that i can improve the PR

File: src/dem.jl [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+            for b in op.bits
+                frame.measurements[f, b] ⊻= true
+            end
+        end
+    end
+    return frame
+end
+
+# The reference trajectory of a Register is noiseless, so the mechanism never fires there.
+apply!(r::Register, ::DetectorError) = r
```
Comment (Krastanov) [inline src/dem.jl]: I am not quite sure I agree with this decision. Is there anywhere else in the codebase that we talk about reference trajectories on a Register? If not, I think it is best to just delete these methods, as they would just cause more confusion.

File: src/dem.jl [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+"""
+struct DeclareMeasurementBits <: AbstractOperation
+    nbits::Int
+end
+
+apply!(frame::PauliFrame, ::DeclareMeasurementBits) = frame
+apply!(r::Register, ::DeclareMeasurementBits) = r
+
+affectedqubits(::DeclareMeasurementBits) = ()
+affectedbits(op::DeclareMeasurementBits) = (op.nbits,)
```
Comment (Krastanov) [inline src/dem.jl:58]: Why is this necessary? Is it used anywhere?

File: src/dem.jl [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+See also: [`detector_error_model_circuit`](@ref), [`parse_detector_error_model`](@ref).
+"""
+read_detector_error_model(path::AbstractString) = parse_detector_error_model(read(path, String))
+
+"""
+$(TYPEDSIGNATURES)
+
+Alias for [`read_detector_error_model`](@ref).
+"""
+detector_error_model_circuit(path::AbstractString) = read_detector_error_model(path)
```
Comment (Krastanov) [inline src/dem.jl]: Why is this alias created? I do not understand how it is helpful.

File: src/dem.jl [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+After sampling, `measurements(frame)` is a `trajectories × (num_detectors +
+num_observables)` `Bool` matrix. The first `num_detectors` columns are the
+detector outcomes `D0, D1, …` and the rest are the observable outcomes
+`L0, L1, …`. As always for Pauli frames, these are relative to the noiseless
+reference run.
+
+See also: [`read_detector_error_model`](@ref).
+"""
+struct DetectorErrorModelCircuit <: AbstractVector{AbstractOperation}
+    circuit::Vector{AbstractOperation}
```
Comment (Krastanov) [inline src/dem.jl]: Should this be abstract operation vector? It seems it can be much more concrete.

In what way is this structure better than just having a list of DetectorErrors?

**Review by Krastanov — COMMENTED** [review]
I gave this a quick first skim. I have a few questions below before I can do a full review. Thanks for looking into this.

File: src/dem.jl (hunk 1 above)
Comment (GrovyleX) [reply src/dem.jl]: Agreed, no precedent for reference-trajectory semantics on Register and DEM circuits only ever run through the frame-path pftrajectories(circuit). Removed both Register methods.

File: src/dem.jl (hunk 2 above)
Comment (GrovyleX) [reply src/dem.jl:58]: It's load-bearing. pftrajectories sizes the measurement buffer from maximum(affectedbits(...)) over the circuit, so a detector/observable that's declared but never appears in an error line would have its column dropped and the output width would disagree with Stim. read_detector_error_model appends one DeclareMeasurementBits(num_detectors+num_observables) to reserve the full width - test_dem.jl covers exactly this (detector D3 with only error D0 ⇒ still 4 columns). Happy to rename if DeclareMeasurementBits isn't the clearest name.

File: src/dem.jl (hunk 3 above)
Comment (GrovyleX) [reply src/dem.jl]: No god reason - the issue listed two candidate names and I kept both by mistake. Dropped the alias, keeping read_detector_error_model.

File: src/dem.jl (hunk 4 above)
Comment (GrovyleX) [reply src/dem.jl]: Fixed the element type - it's only DetectorError + the single DeclareMeasurementBits, so it's now Vector{Union{DetectorError,DeclareMeasurementBits}} (compactify handles fast dispatch in the hot loop).
  The wrapper exists to carry num_detectors/num_observables so callers can split measurements(frames) into detector vs observable columns - the documented output dims the issue asks for. If you'd rather I return a plain vector and surface the counts another way, I'll switch - your call.

**Comment by GrovyleX** [comment]
I chekc the CI.. the only red check is Test QECCore … windows-latest x64 t=1, which errors inside lib/QECCore/test/runtests.jl (1 of 45446, Windows only). This PR doesn't touch lib/QECCore/, and the job tests QECCore against released QuantumClifford, so it looks like a pre-existing flake on the Windows runner rather than something from this change. [If confirmed:] the same test errors on master's Windows job. Could you re-run it? Happy to look closer if you think it's related.
